Added support for proxies

abrgr opened the pull request:
- Feature - Added support for proxying websocket connections

This should address https://github.com/aws/amazon-ecs-agent/issues/23.

euank commented on agent/wsclient/client.go:
```diff
@@ -285,3 +279,72 @@ func (cs *ClientServerImpl) handleMessage(data []byte) {
+		// directly connect
+		log.Info("Creating poll dialer", "host", parsedURL.Host)
+		return tls.DialWithDialer(timeoutDialer, "tcp", targetHost, &tlsConfig)
+	} else {
```
I'd rather not have an else and save a level of indentation here.

euank commented on agent/wsclient/client.go:
```diff
@@ -285,3 +279,72 @@ func (cs *ClientServerImpl) handleMessage(data []byte) {
+			return nil, err
+		}
+		if resp.StatusCode != 200 {
+			f := strings.SplitN(resp.Status, " ", 2)
```
I'd rather use `strconv.Itoa(resp.StatusCode)` than splitting Status. I'm not sure if there's ever a case where Status won't be well formed here, but I also think that the strconv is easier to read anyways.

euank commented on agent/wsclient/client.go:
```diff
@@ -285,3 +279,72 @@ func (cs *ClientServerImpl) handleMessage(data []byte) {
 	}
 }
+
+func (cs *ClientServerImpl) connectWithProxyIfNeeded(parsedURL *url.URL, request *http.Request) (*tls.Conn, error) {
```
I'd rather have this be named something like `websocketConn` and have it return a net.Conn interface. Minor nit though.

euank commented:
Overall, this looks great! I did have a couple minor comments above.

The other thing I'd really like is a quick bit of information on usage here.
Information like "this should work with an HTTP Proxy such as Squid (is that correct?) and can be used by setting NO_PROXY=169.254.169.254 and HTTP_PROXY ..."
Here or as a GoDoc comment would be fine.
We can do the documentation if you'd prefer, but anything you want to contribute there would be awesome!

Oh, also, can you verify that this is fine to take under Apache 2.0?

abrgr commented on agent/wsclient/client.go:
```diff
@@ -285,3 +280,76 @@ func (cs *ClientServerImpl) handleMessage(data []byte) {
+		return tls.DialWithDialer(timeoutDialer, "tcp", targetHost, &tlsConfig)
+	}
+
+	// connect via proxy
```
@euank - Removed the redundant else.

abrgr commented on agent/wsclient/client.go:
```diff
@@ -285,3 +280,76 @@ func (cs *ClientServerImpl) handleMessage(data []byte) {
+	}
+	if resp.StatusCode != 200 {
+		plainConn.Close()
+		return nil, errors.New(strconv.Itoa(resp.StatusCode))
```
Switched from Split to Itoa.  Just a note: this is now creating an error with the status code, not the reason (you get "401" instead of "Unauthorized").

euank replied:
Oops, I thought you were trying to get the code for some reason. My bad.

There's `http.StatusText` which you can pass Status into to preserve the previous behavior. Sorry for the bit of confusion there.

abrgr replied:
Ah!  That makes more sense.  Updated this to use ```.Status``` as you suggest.

abrgr commented on agent/wsclient/client.go:
```diff
@@ -285,3 +280,76 @@ func (cs *ClientServerImpl) handleMessage(data []byte) {
+// (may be listed in /etc/ecs/ecs.config if using ecs-init):
+// HTTP_PROXY=http://<your-proxy>/ # HTTPS_PROXY may be set instead or additionally
+// NO_PROXY=169.254.169.254,/var/run/docker.sock # Directly connect to metadata service and docker socket
+func (cs *ClientServerImpl) websocketConn(parsedURL *url.URL, request *http.Request) (net.Conn, error) {
```
Renamed this method to ```websocketConn``` and it now returns a net.Conn.

abrgr commented:
@euank - Thanks so much for the prompt review.  I believe I've addressed all the issues you've raised.  This is provided to you under Apache 2.0.

Thanks a lot for the great work.

euank commented:
Thanks for addressing my comments. Awesome stuff!

Before merging, I do want to get some testing around this, but I don't have definite timeframe for that.
My initial thought is to add a functional test that spins up a proxy container and runs the agent through it and then either uses the access logs or adds iptable rules to the agent container to verify it was proxied.

If you want to have a go at that bit as well, you're welcome to, but if not we'd be happy to do so and then merge this in.
     
Thanks again,
Euan

abrgr commented:
I totally get the interest in adding some testing.  I don't have the time for that just yet, but I will certainly come back to this in (hopefully) a few weeks and add it if you haven't gotten to it yet.

jdamick commented:
+1 - this is blocking me as awell

euank commented:
I've taken the time to do some manual testing (one EC2 instance running a default-configured squid3 proxy with internet, one EC2 instance with no public IP configuration / nat configuration running the Agent) and everything looked good to me.

I also wrote a functional test to ensure there won't be regressions in the future, which I'll open as a separate pull request shortly.

Thank you so much for adding this feature and your work here!

Apologies for the time it took to get this through; I'd like to think this isn't representative (there was some re:Invent and vacation in there :grimacing:).

Anyways, without further ado, merging! This should be in the next agent release.

Thanks again,
Euan

abrgr commented:
@euank Thanks so much for merging this and for the openness to outside contributions.  In the scheme of things, this got in pretty quickly--awesome!
